src/lib/logs/log_forwarder.py

```diff
@@ -51,11 +54,18 @@ async def run_logs(logging_context: LoggingContext, instance_metadata: InstanceM
 
     # initiate http client for both GCP and DT
     async with init_gcp_client_session() as gcp_session:
-        gcp_client = GCPClient(await create_token(logging_context, gcp_session))
+        token = await create_token(logging_context, gcp_session)
+        if token is None:
+            raise Exception("Cannot start pubsub streaming pull - 'Failed to fetch token")
+        gcp_client = GCPClient(token)
     dynatrace_client = DynatraceClient()
 
+    dt_semaphore = asyncio.Semaphore(NUMBER_OF_CONCURRENT_PUSH_COROUTINES)
```

Comment:
You want to avoid creating semaphore in each loop cycle therefore creating the semaphore here?